build fail looks real
```
======================================================================
ERROR: test suite for <class 'corehq.apps.ota.tests.test_utils.RestorePermissionsTest'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/nose/suite.py", line 209, in run
    self.setUp()
  File "/vendor/lib/python2.7/site-packages/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/vendor/lib/python2.7/site-packages/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/vendor/lib/python2.7/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/mnt/commcare-hq-ro/corehq/apps/ota/tests/test_utils.py", line 71, in setUpClass
    cls.restrict_user_to_location(cls.commcare_user)
AttributeError: type object 'RestorePermissionsTest' has no attribute 'restrict_user_to_location'
```

--- a/corehq/apps/cloudcare/views.py
+++ b/corehq/apps/cloudcare/views.py
@@ -44,6 +44,7 @@ from corehq.apps.app_manager.models import Application, ApplicationBase, RemoteA
 from corehq.apps.app_manager.suite_xml.sections.details import get_instances_for_module
 from corehq.apps.app_manager.suite_xml.sections.entries import EntriesHelper
 from corehq.apps.app_manager.util import get_cloudcare_session_data
+from corehq.apps.locations.permissions import location_safe
 from corehq.apps.cloudcare.api import (
     api_closed_to_status,
     CaseAPIResult,
@@ -233,6 +234,7 @@ class CloudcareMain(View):
             return render(request, "cloudcare/cloudcare_home.html", context)
 
 
+@location_safe
 class FormplayerMain(View):
 
     preview = False

--- a/corehq/apps/locations/permissions.py
+++ b/corehq/apps/locations/permissions.py
@@ -319,6 +319,19 @@ def user_can_access_any_location_id(domain, user, location_ids):
             .exists())
 
 
+def user_can_access_other_user(domain, user, other_user):
+    if user.has_permission(domain, 'access_all_locations'):
+        return True
+
+    accessible_location_ids = (SQLLocation.objects .accessible_to_user(domain, user) .values_list('id', flat=True))
+
+    other_user_locations = other_user.sql_locations
+    return any(map(
+        lambda location_id: location_id in accessible_location_ids,
+        other_user_locations.values_list('id', flat=True) if other_user_locations else []
+    ))
+
+
 def can_edit_location(view_fn):
     """
     Decorator controlling a user's access to a specific location.

--- a/corehq/apps/ota/exceptions.py
+++ b/corehq/apps/ota/exceptions.py
@@ -4,3 +4,7 @@ class PrimeRestoreException(Exception):
 
 class PrimeRestoreUserException(PrimeRestoreException):
     pass
+
+
+class RestorePermissionDenied(Exception):
+    """ Raised when the user is not permitted to restore """

--- a/corehq/apps/ota/tests/test_utils.py
+++ b/corehq/apps/ota/tests/test_utils.py
@@ -1,21 +1,31 @@
 from django.test import TestCase
 
 from corehq.apps.users.models import WebUser, CommCareUser
+from corehq.apps.users.dbaccessors.all_commcare_users import delete_all_users
 from corehq.apps.users.util import format_username
 from corehq.apps.domain.models import Domain
+from corehq.apps.locations.tests.util import LocationHierarchyTestCase
 
 from corehq.apps.ota.utils import is_permitted_to_restore, get_restore_user
 
 
-class RestorePermissionsTest(TestCase):
+class RestorePermissionsTest(LocationHierarchyTestCase):
     domain = 'goats'
     other_domain = 'sheep'
+    location_type_names = ['country', 'state']
+    location_structure = [
+        ('usa', [
+            ('ma', []),
+        ]),
+        ('canada', [
+            ('montreal', []),
+        ]),
+    ]
 
     @classmethod
     def setUpClass(cls):
         super(RestorePermissionsTest, cls).setUpClass()
-        cls.project = Domain(name=cls.domain)
-        cls.project.save()
+        delete_all_users()
 
         cls.other_project = Domain(name=cls.other_domain)
         cls.other_project.save()
@@ -37,14 +47,27 @@ class RestorePermissionsTest(TestCase):
             domain=cls.domain,
             password='***',
         )
+        cls.location_user = CommCareUser.create(
+            username=format_username('location', cls.domain),
+            domain=cls.domain,
+            password='***',
+        )
+        cls.wrong_location_user = CommCareUser.create(
+            username=format_username('wrong-location', cls.domain),
+            domain=cls.domain,
+            password='***',
+        )
+
+        cls.commcare_user.set_location(cls.locations['usa'])
+        cls.location_user.set_location(cls.locations['ma'])
+        cls.wrong_location_user.set_location(cls.locations['montreal'])
+
+        cls.restrict_user_to_location(cls.commcare_user)
 
     @classmethod
     def tearDownClass(cls):
         super(RestorePermissionsTest, cls).tearDownClass()
-        cls.web_user.delete()
-        cls.commcare_user.delete()
-        cls.super_user.delete()
-        cls.project.delete()
+        delete_all_users()
         cls.other_project.delete()
 
     def test_commcare_user_permitted(self):
@@ -141,6 +164,25 @@ class RestorePermissionsTest(TestCase):
         self.assertTrue(is_permitted)
         self.assertIsNone(message)
 
+    def test_commcare_user_as_user_in_location(self):
+        is_permitted, message = is_permitted_to_restore(
+            self.domain,
+            self.commcare_user,
+            u'{}@{}'.format(self.location_user.raw_username, self.domain),
+            True,
+        )
+        self.assertTrue(is_permitted)
+        self.assertIsNone(message)
+
+        is_permitted, message = is_permitted_to_restore(
+            self.domain,
+            self.commcare_user,
+            u'{}@{}'.format(self.wrong_location_user.raw_username, self.domain),
+            True,
+        )
+        self.assertFalse(is_permitted)
+        self.assertIsNotNone(message)
+
 
 class GetRestoreUserTest(TestCase):
 
@@ -163,6 +205,11 @@ class GetRestoreUserTest(TestCase):
         )
         cls.commcare_user = CommCareUser.create(
             username=format_username('jane', cls.domain),
+            domain=cls.domain,
+            password='***',
+        )
+        cls.other_commcare_user = CommCareUser.create(
+            username=format_username('john', cls.domain),
             domain=cls.domain,
             password='***',
         )
@@ -224,3 +271,11 @@ class GetRestoreUserTest(TestCase):
                 '{}@wrong-domain'.format(self.commcare_user.raw_username, self.domain)
             )
         )
+
+    def test_get_restore_user_as_user_for_commcare_user(self):
+        user = get_restore_user(
+            self.domain,
+            self.commcare_user,
+            '{}@{}'.format(self.other_commcare_user.raw_username, self.domain)
+        )
+        self.assertEquals(user.user_id, self.other_commcare_user._id)

--- a/corehq/apps/ota/utils.py
+++ b/corehq/apps/ota/utils.py
@@ -11,6 +11,8 @@ from .models import DemoUserRestore
 from dimagi.utils.web import json_response
 from corehq.apps.domain.auth import get_username_and_password_from_request, determine_authtype_from_request
 from corehq.apps.users.decorators import ensure_active_user_by_username
+from corehq.apps.locations.permissions import user_can_access_other_user
+
 
 def turn_off_demo_mode(commcare_user):
     """
@@ -104,9 +106,7 @@ def is_permitted_to_restore(domain, couch_user, as_user, has_data_cleanup_privel
         message = u"{} was not in the domain {}".format(couch_user.username, domain)
     elif couch_user.is_web_user() and domain not in couch_user.domains and not couch_user.is_superuser:
         message = u"{} was not in the domain {}".format(couch_user.username, domain)
-    elif (couch_user.is_web_user() and
-            couch_user.is_member_of(domain) and
-            as_user is not None):
+    elif (couch_user.is_member_of(domain) and as_user is not None):
         if not has_data_cleanup_privelege and not couch_user.is_superuser:
             message = u"{} does not have permissions to restore as {}".format(
                 couch_user.username,
@@ -118,7 +118,6 @@ def is_permitted_to_restore(domain, couch_user, as_user, has_data_cleanup_privel
             user_domain = as_user.split('@')[1]
         except IndexError:
             message = u"Invalid to restore user {}. Format is <user>@<domain>".format(as_user)
-
         else:
             if user_domain != domain:
                 # In this case we may be dealing with a WebUser
@@ -127,6 +126,14 @@ def is_permitted_to_restore(domain, couch_user, as_user, has_data_cleanup_privel
                     message = None
                 else:
                     message = u"{} was not in the domain {}".format(username, domain)
+
+        # If it's a commcare user ensure that we check location restrictions
+        if couch_user.is_commcare_user():
+            as_user_obj = CommCareUser.get_by_username('{}.commcarehq.org'.format(as_user))
+            if not as_user_obj:
+                message = u'Invalid restore user {}'.format(as_user)
+            elif not user_can_access_other_user(domain, couch_user, as_user_obj):
+                message = u'Restore user {} not in allowed locations'.format(as_user)
 
     return message is None, message
 
@@ -142,9 +149,7 @@ def get_restore_user(domain, couch_user, as_user):
     :returns: An instance of OTARestoreUser
     """
     restore_user = None
-    if couch_user.is_commcare_user():
-        restore_user = couch_user.to_ota_restore_user()
-    elif (couch_user.is_web_user() and as_user is not None):
+    if as_user is not None:
         if '@' in as_user:
             _, user_domain = as_user.split('@')
         else:
@@ -163,7 +168,8 @@ def get_restore_user(domain, couch_user, as_user):
             elif not user.is_member_of(domain):
                 return None
             restore_user = user.to_ota_restore_user(domain)
-
+    elif couch_user.is_commcare_user():
+        restore_user = couch_user.to_ota_restore_user()
     elif couch_user.is_web_user():
         restore_user = couch_user.to_ota_restore_user(domain)
 

--- a/corehq/apps/ota/utils.py
+++ b/corehq/apps/ota/utils.py
@@ -6,12 +6,13 @@ from dimagi.utils.logging import notify_exception
 
 from corehq.apps.users.models import CommCareUser, WebUser
 
-from .models import DemoUserRestore
-
 from dimagi.utils.web import json_response
 from corehq.apps.domain.auth import get_username_and_password_from_request, determine_authtype_from_request
 from corehq.apps.users.decorators import ensure_active_user_by_username
 from corehq.apps.locations.permissions import user_can_access_other_user
+
+from .models import DemoUserRestore
+from .exceptions import RestorePermissionDenied
 
 
 def turn_off_demo_mode(commcare_user):
@@ -89,7 +90,45 @@ def demo_restore_date_created(commcare_user):
             return restore.timestamp_created
 
 
-def is_permitted_to_restore(domain, couch_user, as_user, has_data_cleanup_privelege):
+def _ensure_valid_domain(domain, couch_user):
+    if not couch_user.is_member_of(domain):
+        raise RestorePermissionDenied(u"{} was not in the domain {}".format(couch_user.username, domain))
+
+
+def _ensure_cleanup_permission(domain, couch_user, as_user, has_data_cleanup_privilege):
+    if not has_data_cleanup_privilege and not couch_user.is_superuser:
+        raise RestorePermissionDenied(u"{} does not have permissions to restore as {}".format(
+            couch_user.username,
+            as_user,
+        ))
+
+
+def _ensure_valid_restore_as_user(domain, couch_user, as_user):
+    try:
+        username = as_user.split('@')[0]
+        user_domain = as_user.split('@')[1]
+    except IndexError:
+        raise RestorePermissionDenied(
+            u"Invalid to restore user {}. Format is <user>@<domain>".format(as_user)
+        )
+    else:
+        if user_domain != domain:
+            # In this case we may be dealing with a WebUser
+            user = WebUser.get_by_username(as_user)
+            if not user or not user.is_member_of(domain):
+                raise RestorePermissionDenied(u"{} was not in the domain {}".format(username, domain))
+
+
+def _ensure_accessible_location(domain, couch_user, as_user):
+    if couch_user.is_commcare_user():
+        as_user_obj = CommCareUser.get_by_username('{}.commcarehq.org'.format(as_user))
+        if not as_user_obj:
+            raise RestorePermissionDenied(u'Invalid restore user {}'.format(as_user))
+        elif not user_can_access_other_user(domain, couch_user, as_user_obj):
+            raise RestorePermissionDenied(u'Restore user {} not in allowed locations'.format(as_user))
+
+
+def is_permitted_to_restore(domain, couch_user, as_user, has_data_cleanup_privilege):
     """
     This function determines if the couch_user is permitted to restore
     for the domain and/or as_user
@@ -101,41 +140,16 @@ def is_permitted_to_restore(domain, couch_user, as_user, has_data_cleanup_privel
     :returns: a tuple - first a boolean if the user is permitted,
         secondly a message explaining why a user was rejected if not permitted
     """
-    message = None
-    if couch_user.is_commcare_user() and domain != couch_user.domain:
-        message = u"{} was not in the domain {}".format(couch_user.username, domain)
-    elif couch_user.is_web_user() and domain not in couch_user.domains and not couch_user.is_superuser:
-        message = u"{} was not in the domain {}".format(couch_user.username, domain)
-    elif (couch_user.is_member_of(domain) and as_user is not None):
-        if not has_data_cleanup_privelege and not couch_user.is_superuser:
-            message = u"{} does not have permissions to restore as {}".format(
-                couch_user.username,
-                as_user,
-            )
-
-        try:
-            username = as_user.split('@')[0]
-            user_domain = as_user.split('@')[1]
-        except IndexError:
-            message = u"Invalid to restore user {}. Format is <user>@<domain>".format(as_user)
-        else:
-            if user_domain != domain:
-                # In this case we may be dealing with a WebUser
-                user = WebUser.get_by_username(as_user)
-                if user and user.is_member_of(domain):
-                    message = None
-                else:
-                    message = u"{} was not in the domain {}".format(username, domain)
-
-        # If it's a commcare user ensure that we check location restrictions
-        if couch_user.is_commcare_user():
-            as_user_obj = CommCareUser.get_by_username('{}.commcarehq.org'.format(as_user))
-            if not as_user_obj:
-                message = u'Invalid restore user {}'.format(as_user)
-            elif not user_can_access_other_user(domain, couch_user, as_user_obj):
-                message = u'Restore user {} not in allowed locations'.format(as_user)
-
-    return message is None, message
+    try:
+        _ensure_valid_domain(domain, couch_user)
+        if as_user is not None:
+            _ensure_cleanup_permission(domain, couch_user, as_user, has_data_cleanup_privilege)
+            _ensure_valid_restore_as_user(domain, couch_user, as_user)
+            _ensure_accessible_location(domain, couch_user, as_user)
+    except RestorePermissionDenied as e:
+        return False, unicode(e)
+    else:
+        return True, None
 
 
 def get_restore_user(domain, couch_user, as_user):

--- a/corehq/apps/ota/views.py
+++ b/corehq/apps/ota/views.py
@@ -25,6 +25,7 @@ from corehq.apps.hqwebapp.views import BaseSectionPageView
 from corehq.apps.ota.forms import PrimeRestoreCacheForm, AdvancedPrimeRestoreCacheForm
 from corehq.apps.ota.tasks import queue_prime_restore
 from corehq.apps.users.models import CouchUser, CommCareUser
+from corehq.apps.locations.permissions import location_safe
 from corehq.form_processor.exceptions import CaseNotFound
 from corehq.pillows.mappings.case_search_mapping import CASE_SEARCH_MAX_RESULTS
 from corehq.tabs.tabclasses import ProjectSettingsTab
@@ -37,6 +38,7 @@ from soil import MultipleTaskDownload
 from .utils import demo_user_restore_response, get_restore_user, is_permitted_to_restore, handle_401_response
 
 
+@location_safe
 @json_error
 @handle_401_response
 @login_or_digest_or_basic_or_apikey()

